marimo/_save/ast.py

```diff
@@ -106,31 +112,69 @@ def generic_visit(self, node: ast.AST) -> tuple[ast.Module, ast.Module]:  #  typ
-                        previous.body  # type: ignore[arg-type]
-                    )
-                except BlockException:
-                    on_line.append(previous)
+                on_line.append(previous)
+                bodies = (
+                    [previous.body]
+                    if isinstance(previous, ast.With)
+                    else [previous.body, previous.orelse]
+                )
+                for body in bodies:
+                    try:
+                        # Recursion by referring the the containing block also
+                        # captures the case where the target line number was not
+                        # exactly hit.
+                        return ExtractWithBlock(
+                            self.target_line
+                        ).generic_visit(
+                            body  # type: ignore[arg-type]
+                        )
+                    except BlockException:
+                        pass
+
             else:
                 raise BlockException(
                     "persistent_cache cannot be invoked within a block "
                     "(try moving the block within the persistent_cache scope)."
                 )
         # Intentionally not elif (on_line can be added in previous block)
         if len(on_line) == 1:
-            assert isinstance(on_line[0], ast.With), "Unexpected block."
+            if parent and not isinstance(on_line[0], ast.With):
+                raise BlockException("Detected line is not a With statement.")
+            if not isinstance(on_line[0], ast.With):
+                raise BlockException(
+                    "Saving on a shared line may lead to unexpected behavior. "
+                    "Please format your code, and or reduce nesting."
+                )
             return clean_to_modules(pre_block, on_line[0])
         # It should be possible to relate the lines with the AST,
         # but reduce potential bugs by just throwing an error.
         raise BlockException(
-            "Saving on a shared line may lead to unexpected behavior."
+            "Unable to determine structure your call. Please"
+            " report this to github:marimo-team/marimo/issues"
         )
 
 
+class MangleArguments(ast.NodeTransformer):
+    """Mangles arguments names to prevent shadowing issues in analysis."""
+
+    def __init__(
+        self, prefix: str, args: set[str], *arg: Any, **kwargs: Any
+    ) -> None:
+        super().__init__(*arg, **kwargs)
+        self.prefix = prefix
+        self.args = args
+
+    def visit_Name(self, node: ast.Name) -> ast.Name:
+        if node.id in self.args:
+            node.id = f"{self.prefix}{node.id}"
+        return node
+
+    def generic_visit(self, node: ast.AST) -> ast.AST:
+        if hasattr(node, "name") and node.name in self.args:
+            node.name = f"{self.prefix}{node.name}"
+        return super().generic_visit(node)
```

Comment:
This is for ast nodes that are essentially names but aren't `Name` nodes?

Reply:
So there are edge cases where the name is an attribute, and not a Name node (e.g. a FunctionDef name), but they are primarily in definitions and not relevant to scope. I did a dump of some functions to see where names popped up, and may have been overzealous since I don't think it's possible to get an erroneous ref